### PR TITLE
Allow Corretto 8 to be built with Xcode 6 + SDK 10.10

### PR DESCRIPTION
**Pull request by ziyiluo**
This is a squashed commit of changes made to build on Xcode 6 and macOS
SDK 10.10. This makes Corretto 8 build artifacts meet the minimum SDK
requirement for app notarization.

Additional de-optimizations were added to two JVM files `unsafe.o` and `metaspaceShared.o` to overcome clang optimization bugs.

After this change, Corretto 8 can still be built with Xcode 4 + gcc toolchain. In my local experiment, it can also be built with Xcode 8 with trivial configuration changes.

### CODE REVIEW

File: src/hotspot/make/bsd/makefiles/gcc.make [hunk 1]
```diff
@@ -317,6 +319,14 @@ ifeq ($(USE_CLANG), true)
   ifeq ($(shell expr $(CC_VER_MAJOR) = 4 \& $(CC_VER_MINOR) = 2), 1)
     OPT_CFLAGS/loopTransform.o += $(OPT_CFLAGS/NOOPT)
     OPT_CFLAGS/unsafe.o += -O1
+    OPT_CFLAGS/metaspaceShared.o += $(OPT_CFLAGS/NOOPT)
+  endif
+
+  # We suspect that this bug was not addressed in Clang 6. Deopt unsafe.o
+  ifeq ($(shell expr $(CC_VER_MAJOR) = 6), 1)
```
Comment (navyxliu) [inline src/hotspot/make/bsd/makefiles/gcc.make]: >= 6?

File: src/common/autoconf/generated-configure.sh [hunk 2]
```diff
@@ -1895,8 +1895,8 @@ Optional Packages:
   --with-toolchain-path   prepend these directories when searching for
                           toolchain binaries (compilers etc)
   --with-extra-path       prepend these directories to the default path
-  --with-xcode-path       explicit path to Xcode 4 (generally for building on
-                          10.9 and later)
+  --with-xcode-path       explicit path to Xcode (Xcode 4 is supported;
```
Comment (navyxliu) [inline src/common/autoconf/generated-configure.sh]: I think you intend to support both xcode 4 & 6.

File: src/hotspot/make/bsd/makefiles/gcc.make [hunk 3]
```diff
@@ -317,6 +319,14 @@ ifeq ($(USE_CLANG), true)
...
+    OPT_CFLAGS/loopTransform.o += $(OPT_CFLAGS/NOOPT)
+    OPT_CFLAGS/unsafe.o += -O1
+    OPT_CFLAGS/metaspaceShared.o += $(OPT_CFLAGS/NOOPT)
```
Comment (navyxliu) [inline src/hotspot/make/bsd/makefiles/gcc.make]: Could you leave paper trail here why metaspaceShared.cpp also needs special handle?
Further, please switch this line with 328.

File: src/hotspot/make/bsd/makefiles/gcc.make [hunk 4]
```diff
@@ -363,7 +373,15 @@ ASFLAGS += -x assembler-with-cpp
 # Linker flags
 
 # statically link libstdc++.so, work with gcc but ignored by g++
-STATIC_STDCXX = -Wl,-Bstatic -lstdc++ -Wl,-Bdynamic
```
Comment (navyxliu) [inline src/hotspot/make/bsd/makefiles/gcc.make]: I don't think you have to check OS_VENODR  here.  Basically, it's clang/gcc thing. 
g++ uses libstdc++. clang++ links libc++.  It should stick with libc++ even on linux platform.

File: src/common/autoconf/generated-configure.sh [hunk 5]
```diff
@@ -41438,6 +41471,20 @@ $as_echo "$ac_cv_c_bigendian" >&6; }
         C_O_FLAG_NORM="-O2"
         C_O_FLAG_NONE="-O0"
       fi
+    elif test "x$TOOLCHAIN_TYPE" = xclang; then
+      if test "x$OPENJDK_TARGET_OS" = xmacosx; then
+        # On MacOSX we optimize for size, something
+        # we should do for all platforms?
+        C_O_FLAG_HIGHEST="-Os"
```
Comment (navyxliu) [inline src/common/autoconf/generated-configure.sh]: This is weird. why does it emphasize on code size for MacOS?

File: src/hotspot/make/bsd/makefiles/gcc.make (hunk 1 above)
Comment (ziyiluo) [reply src/hotspot/make/bsd/makefiles/gcc.make]: I cannot confirm if this bug exists in newer clangs (therefore I prefer not to introduce potential unnecessary de-opts). This commit reflects experiments and tests I ran exclusively against Xcode 6 (w/ clang 6). We may change this later with more test data with newer Xcodes.

File: src/common/autoconf/generated-configure.sh (hunk 5 above)
Comment (ziyiluo) [reply src/common/autoconf/generated-configure.sh]: This is the same flag as using gcc on macOS. But I can run an experiment to see the diff in size by switching this flag to `-O3`.

Comment (navyxliu) [reply src/common/autoconf/generated-configure.sh]: the tip is also same. 
https://hg.openjdk.java.net/jdk/jdk/file/7ef41e83066b/make/autoconf/flags-cflags.m4#l302

**Pull request merged**